build: add bazel build files to most packages

alexeagle opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview 41c947b at https://pr20768-41c947b.ngbuilds.io/.

mary-poppins commented:
You can preview 85de38a at https://pr20768-85de38a.ngbuilds.io/.

mary-poppins commented:
You can preview c7a593c at https://pr20768-c7a593c.ngbuilds.io/.

mary-poppins commented:
You can preview ca56d91 at https://pr20768-ca56d91.ngbuilds.io/.

mary-poppins commented:
You can preview 9f8d433 at https://pr20768-9f8d433.ngbuilds.io/.

mary-poppins commented:
You can preview f496a7e at https://pr20768-f496a7e.ngbuilds.io/.

mary-poppins commented:
You can preview 624f0c9 at https://pr20768-624f0c9.ngbuilds.io/.

IgorMinar commented on package.json:
```diff
@@ -24,15 +24,16 @@
     "core-js": "^2.4.1",
     "reflect-metadata": "^0.1.3",
-    "rxjs": "^5.5.2",
+    "rxjs": "git+https://github.com/gregmagolan/rxjs.git#rxjs-bazel-dist-2",
```
Are these changes being upstreamed? Can you link to the PR from the commit message?

alexeagle replied:
yes, https://github.com/ReactiveX/rxjs/pull/3131

alexeagle replied:
Added https://github.com/ReactiveX/rxjs/pull/3131 to commit msg

IgorMinar commented on .circleci/config.yml:
```diff
@@ -42,7 +42,7 @@ jobs:
       - run: bazel run @yarn//:yarn
       - run: bazel build packages/...
-      - run: bazel test @angular//...
+      - run: bazel test packages/... @angular/...
```
what is `@angular/...` that is not under packages/... ?

alexeagle replied:
@angular is the name of the workspace that lives in packages/bazel. That's how users can
`load("@angular//:index.bzl", "ng_module")`

The @angular_src workspace lives in the root dir.

The outer workspace doesn't see anything inside the nested workspace.

IgorMinar commented on packages/BUILD.bazel:
```diff
@@ -1 +1,20 @@
+ts_library(
+    name = "types",
+    srcs = glob(["*.ts"]),
+    tsconfig = "tsconfig.json",
```
so unlike in g3, externally we allow tsconfig.json to be specified?

IgorMinar replied:
I'm not opposed to it. I think that's the only way externally, I'm just wondering if this is intentional.

alexeagle replied:
correct.
Before landing this PR I'd like to fix https://github.com/bazelbuild/rules_typescript/issues/93 so we don't have to specify this everywhere.
Note that the BUILD files only ever refer to one tsconfig file, so after fully switching to Bazel we can delete the others. The information they contain is just the dependency graph.

IgorMinar commented on packages/animations/browser/testing/BUILD.bazel:
```diff
@@ -0,0 +1,15 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
"/testing" secondary entry point is not `testonly` - it's prod code that we ship to npm.

tests are only those files that end with `_spec.ts` or `.spec.ts` typically in the `test/` directory. eventually, I'd prefer to mix the src and test code so that we don't have two parallel file trees but that's not what you should be trying to do here.

alexeagle replied:
That's not what `testonly` means. It's applied both to tests, and to test-support code.

`testonly` is a marker that prevents a non-testonly target from depending on a `testonly` one - it taints the build graph and bazel enforces this.

We can still ship `testonly` packages to NPM. what this does is prevent mistakes like `//packages/common` depending on `//packages/core/testing`, and also when Angular/Bazel users make a source dependency on Angular, they'll get the same constraint so they don't ship testing code.

IgorMinar replied:
I see. that makes sense. thanks for explaining!

IgorMinar commented on packages/bazel/test/ngc-wrapped/test_support.ts:
```diff
@@ -150,7 +150,7 @@ function makeTempDir(baseDir): string {
 
 export function listFilesRecursive(dir: string, fileList: string[] = []) {
-  fs.readdirSync(dir).map(file => {
+  fs.readdirSync(dir).forEach(file => {
```
is this less surprising (since map returns something that is not used) or why this change?

alexeagle replied:
this change is required.
This doesn't type-check under Bazel, because it violates http://tsetse.info/check-return-value

IgorMinar commented on packages/common/BUILD.bazel:
```diff
@@ -1,15 +1,19 @@
-package(default_visibility=["//visibility:public"])
+package(default_visibility = ["//visibility:public"])
```
finished reviewing up to here. getting of the bus, so I'm posting what I have for now.

alexeagle replied:
:+1:

IgorMinar reviewed (commented):
posting review part 1

IgorMinar commented on packages/common/http/testing/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/common/testing/BUILD.bazel:
```diff
@@ -0,0 +1,16 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

alexeagle replied:
yes!

IgorMinar commented on packages/compiler/test/BUILD.bazel:
```diff
@@ -0,0 +1,22 @@
+
+ts_library(
+    name = "test",
+    testonly = 1,
```
yup :)

IgorMinar commented on packages/compiler/test/BUILD.bazel:
```diff
@@ -0,0 +1,22 @@
+        "//packages/core/testing",
+        "//packages/platform-browser",
+        "//packages/platform-browser-dynamic",
+        "//packages/platform-browser/testing",
```
how did you determine this dep list? trial and error? something else?

alexeagle replied:
Iterated with `ibazel` until the build was green.

We really need tazel, which my intern  is scheduled to write.

These lists are exhaustive to satisfy strict deps. Note that we now have the option to opt-in or opt-out of strict deps. We should discuss.

IgorMinar commented on packages/compiler/testing/BUILD.bazel:
```diff
@@ -0,0 +1,12 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/core/testing/BUILD.bazel:
```diff
@@ -0,0 +1,15 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/core/testing/BUILD.bazel:
```diff
@@ -0,0 +1,15 @@
+    module_name = "@angular/core/testing",
+    tsconfig = "//packages:tsconfig",
+    deps = [
+        "//packages:types",
```
what's this one?

alexeagle replied:
Our random grab-bag of typings in the packages folder, `goog.d.ts` and `es6-subset.d.ts` and so on

IgorMinar commented on packages/http/testing/BUILD.bazel:
```diff
@@ -0,0 +1,16 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/common/BUILD.bazel:
```diff
@@ -1,15 +1,19 @@
+
 load("@build_bazel_rules_typescript//:defs.bzl", "ts_library")
 
 ts_library(
```
shouldn't this be an ng_module rule?

IgorMinar replied:
also elsewhere...

alexeagle replied:
yes, they should, assuming there are any decorators that require summary files. I got some errors doing so, but we can push more on them.

Note that this PR is incomplete - there are probably more packages missing build files. The motivation for this PR is the minimal set of changes needed to compile the `packages/core/test/**/*.ts`

IgorMinar replied:
ah. that was not obvious to me.

IgorMinar commented on packages/platform-browser-dynamic/BUILD.bazel:
```diff
@@ -0,0 +1,23 @@
+        ],
+    ),
+    module_name = "@angular/platform-browser-dynamic",
+    tsconfig = "//packages:tsconfig",
```
is it right that you are reusing this single tsconfig file for all the ts_libraries? we have package specific tsconfigs already, so I'm surprised that you are not using those..

alexeagle replied:
We should not need package-specific tsconfigs anymore. The google model, and the recommendation to all teams, is to have one tsconfig for a monorepo

IgorMinar replied:
interesting. ok

IgorMinar commented on packages/platform-browser-dynamic/testing/BUILD.bazel:
```diff
@@ -0,0 +1,20 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/platform-browser/testing/BUILD.bazel:
```diff
@@ -0,0 +1,16 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/platform-server/testing/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+    deps = [
+        "//packages/core",
+        "//packages/platform-browser-dynamic/testing",
+        "//packages/platform-browser/animations",
```
animations? really?

alexeagle replied:
really! if you want this to compile, that is...

IgorMinar commented on packages/router/testing/BUILD.bazel:
```diff
@@ -0,0 +1,17 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/service-worker/BUILD.bazel:
```diff
@@ -0,0 +1,19 @@
+    module_name = "@angular/service-worker",
+    tsconfig = "//packages:tsconfig",
+    deps = [
+        "//packages/core",
```
I think we now also depend on common. we just merged a PR that does that.

alexeagle replied:
no problem, will catch that in CI

IgorMinar commented on packages/service-worker/worker/testing/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+
+ts_library(
+    name = "testing",
+    testonly = 1,
```
nope

IgorMinar commented on packages/tsconfig-build.json:
```diff
@@ -23,5 +23,9 @@
+  },
+
+  "bazelOptions": {
+    "disableStrictDeps": true
```
can you leave a comment in here to explain why this is turned off?

alexeagle replied:
I think we can turn it on now, I just merged a fix for strictdeps lookup paths

IgorMinar commented on yarn.lock:
```diff
@@ -7156,6 +7161,12 @@ tsscmp@1.0.5:
   version "1.0.5"
   resolved "https://registry.yarnpkg.com/tsscmp/-/tsscmp-1.0.5.tgz#7dc4a33af71581ab4337da91d85ca5427ebd9a97"
 
+tsutils@2.12.1:
```
what is this one here for? I didn't notice it being used directly by any of your changes....

alexeagle replied:
it's a workaround for an issue I spent all day yesterday landing, see https://github.com/bazelbuild/rules_typescript/pull/95

alexeagle replied:
Workaround is still needed to land this, but I'll keep working on the underlying issue now that I understand it better.

IgorMinar reviewed (changes requested):
done with the first full review pass..

mary-poppins commented:
You can preview ab0bec2 at https://pr20768-ab0bec2.ngbuilds.io/.

mary-poppins commented:
You can preview f096e3b at https://pr20768-f096e3b.ngbuilds.io/.

mary-poppins commented:
You can preview 9d0aa64 at https://pr20768-9d0aa64.ngbuilds.io/.

mary-poppins commented:
You can preview 6bb9afa at https://pr20768-6bb9afa.ngbuilds.io/.

alexeagle committed:
build: upgrade bazel rules to latest

Add enough BUILD files to make it possible to
`bazel build packages/core/test`

Also re-format BUILD.bazel files with Buildifier.
Add a CI lint check that they stay formatted.

mary-poppins commented:
You can preview 3d2e6da at https://pr20768-3d2e6da.ngbuilds.io/.

mary-poppins commented:
You can preview 637c543 at https://pr20768-637c543.ngbuilds.io/.

mary-poppins commented:
You can preview 2fd5c7a at https://pr20768-2fd5c7a.ngbuilds.io/.

mary-poppins commented:
You can preview ddfa023 at https://pr20768-ddfa023.ngbuilds.io/.

mary-poppins commented:
You can preview 2261b45 at https://pr20768-2261b45.ngbuilds.io/.

mary-poppins commented:
You can preview b399990 at https://pr20768-b399990.ngbuilds.io/.

mary-poppins commented:
You can preview a04d325 at https://pr20768-a04d325.ngbuilds.io/.

mary-poppins commented:
You can preview 6e3bb07 at https://pr20768-6e3bb07.ngbuilds.io/.

mary-poppins commented:
You can preview 2f9dffe at https://pr20768-2f9dffe.ngbuilds.io/.

mary-poppins commented:
You can preview 93ffac0 at https://pr20768-93ffac0.ngbuilds.io/.

mary-poppins commented:
You can preview 44cfd01 at https://pr20768-44cfd01.ngbuilds.io/.

mary-poppins commented:
You can preview 8a0d73d at https://pr20768-8a0d73d.ngbuilds.io/.

mary-poppins commented:
You can preview 1fc33ce at https://pr20768-1fc33ce.ngbuilds.io/.

mary-poppins commented:
You can preview ed00f63 at https://pr20768-ed00f63.ngbuilds.io/.

mary-poppins commented:
You can preview af27d1b at https://pr20768-af27d1b.ngbuilds.io/.
